File: src/swell_lobster/api/routes/config_env.py
"""
Config routes — .env 环境变量管理

覆盖接口:
- GET  /api/config/env   读取 .env（敏感值脱敏）
- POST /api/config/env   更新 .env 键值（合并，保留注释）
"""

from __future__ import annotations

import logging
import re
from pathlib import Path

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from swell_lobster.config import settings
from swell_lobster.utils.env_utils import parse_env, update_env_content

logger = logging.getLogger(__name__)

router = APIRouter()

# 敏感键名匹配模式
_SENSITIVE = re.compile(r"(TOKEN|SECRET|PASSWORD|KEY|APIKEY)", re.IGNORECASE)
# 合法环境变量键名
_KEY_PATTERN = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")


# ── Helpers ────────────────────────────────────────────────────────────────────


def _env_path() -> Path:
    return Path(settings.project_root) / ".env"


# ── Pydantic models ────────────────────────────────────────────────────────────


class EnvUpdateRequest(BaseModel):
    """更新 .env 的键值对。"""
    entries: dict[str, str]


# ── Routes ─────────────────────────────────────────────────────────────────────


@router.get("/api/config/env")
async def read_env() -> dict:
    """读取 .env 为键值对（敏感值脱敏后返回）。"""
    path = _env_path()
    if not path.exists():
        return {"env": {}, "raw": ""}

    content = path.read_text(encoding="utf-8", errors="replace")
    env = parse_env(content)

    masked = {
        k: (v[:4] + "***" + v[-2:] if len(v) > 6 else "***")
        if _SENSITIVE.search(k) and v
        else v
        for k, v in env.items()
    }
    return {"env": masked, "masked": masked, "raw": ""}


@router.post("/api/config/env")
async def write_env(body: EnvUpdateRequest) -> dict:
    """更新 .env 键值（合并写入，保留注释与原有顺序）。"""
    for key in body.entries:
        if not _KEY_PATTERN.match(key):
            raise HTTPException(status_code=400, detail=f"Invalid env key: {key!r}")

    path = _env_path()
    existing = path.read_text(encoding="utf-8", errors="replace") if path.exists() else ""
    new_content = update_env_content(existing, body.entries)

    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(new_content, encoding="utf-8")
    logger.info("[Env API] Updated .env with %d entries", len(body.entries))
    return {"status": "ok", "updated_keys": list(body.entries.keys())}

File: src/tide-lobster/src/api/routes/configEnv.ts
/**
 * Config routes — .env 环境变量管理
 *
 * Node 后端实现（原 Python 参考已移除）。
 *
 * 接口：
 * - GET  /api/config/env   读取 .env（敏感值脱敏）
 * - POST /api/config/env   更新 .env 键值（合并，保留注释）
 */

import { Hono } from 'hono';
import { readFileSync, writeFileSync, existsSync, mkdirSync } from 'node:fs';
import { resolve, dirname } from 'node:path';
import { settings } from '../../config.js';
import { parseEnv, updateEnvContent } from '../../utils/envUtils.js';

export const configEnvRouter = new Hono();

// 敏感键名匹配
const SENSITIVE = /(TOKEN|SECRET|PASSWORD|KEY|APIKEY)/i;
// 合法环境变量键名
const KEY_PATTERN = /^[A-Za-z_][A-Za-z0-9_]*$/;

function envPath(): string {
  return resolve(settings.projectRoot, '.env');
}

function maskValue(key: string, value: string): string {
  if (SENSITIVE.test(key) && value) {
    return value.length > 6 ? value.slice(0, 4) + '***' + value.slice(-2) : '***';
  }
  return value;
}

// ── GET /api/config/env ────────────────────────────────────────────────────────

configEnvRouter.get('/api/config/env', (c) => {
  const path = envPath();
  if (!existsSync(path)) return c.json({ env: {}, raw: '' });

  const content = readFileSync(path, 'utf-8');
  const env = parseEnv(content);
  const masked = Object.fromEntries(Object.entries(env).map(([k, v]) => [k, maskValue(k, v)]));
  return c.json({ env: masked, masked, raw: '' });
});

// ── POST /api/config/env ───────────────────────────────────────────────────────

configEnvRouter.post('/api/config/env', async (c) => {
  const body = await c.req.json<{ entries: Record<string, string> }>();
  const entries = body.entries ?? {};

  for (const key of Object.keys(entries)) {
    if (!KEY_PATTERN.test(key)) {
      return c.json({ error: `Invalid env key: ${JSON.stringify(key)}` }, 400);
    }
  }

  const path = envPath();
  const existing = existsSync(path) ? readFileSync(path, 'utf-8') : '';
  const newContent = updateEnvContent(existing, entries);

  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, newContent, 'utf-8');

  return c.json({ status: 'ok', updated_keys: Object.keys(entries) });
});
